File: 3. Longest Substring Without Repeating Characters-1.cpp
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    unordered_set<char> uset;

    int i = 0, res = 0;
    for (int j = 0; j < s.size(); ++j)
      if (!uset.count(s[j])) {
        uset.insert(s[j]);
        res = max(res, (int)uset.size());
      } else {
        while (i < j && s[i] != s[j]) {
          uset.erase(s[i]);
          i++;
        }
        i++;
      }

    return res;
  }
};

File: 3. Longest Substring Without Repeating Characters-1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        cnt = Counter()  # char, int
        left = 0
        for right, ch in enumerate(s):
            cnt[ch] += 1
            while cnt[ch] > 1:
                cnt[s[left]] -= 1
                left += 1
            res = max(res, right - left + 1)

        return res
